File: db/threads.go
package db

import (
	"context"

	"github.com/bakape/pg_util"
	"github.com/jackc/pgx/v4"
)

type ThreadInsertParams struct {
	Subject string

	// Must include between 1 and 3 tags
	Tags []string

	PostInsertParamsCommon `db:"-"`
}

// Insert thread and empty post into DB and return the post ID.
//
// authKey is optional, in case this is a migration of a legacy post
func InsertThread(ctx context.Context, p ThreadInsertParams) (
	id uint64, err error,
) {
	err = InTransaction(context.Background(), func(tx pgx.Tx) (err error) {
		q, args := pg_util.BuildInsert(pg_util.InsertOpts{
			Table:  "threads",
			Data:   p,
			Suffix: "returning id",
		})
		err = tx.QueryRow(ctx, q, args...).Scan(&id)
		if err != nil {
			return
		}

		_, err = InsertPost(ctx, tx, OPInsertparams{
			ID: id,
			ReplyInsertParams: ReplyInsertParams{
				Thread:                 id,
				PostInsertParamsCommon: p.PostInsertParamsCommon,
			},
		})
		return
	})
	return
}

// Check, if thread exists in the database
func ThreadExists(ctx context.Context, id uint64) (exists bool, err error) {
	err = db.
		QueryRow(
			ctx,
			`select exists (
				select
				from threads
				where id = $1
			)`,
			id,
		).
		Scan(&exists)
	return
}

// Read feed data for initializing Pulsar as JSON
func GetFeedData() (buf []byte, err error) {
	err = db.
		QueryRow(
			context.Background(),
			`select
				jsonb_agg(
					jsonb_build_object(
						'thread', t.id,
						'recent_posts', coalesce(r.val, '{}'::jsonb),
						'open_posts', coalesce(o.val, '{}'::jsonb)
					)
				)
			from threads t
			left join (
				select
					r.thread,
					jsonb_object_agg(
						r.id,
						to_unix(r.created_on)
					) val
				from posts r
				where r.created_on > now() - interval '16 minutes'
				group by r.thread
			) r on r.thread = t.id
			left join (
				select
					o.thread,
					jsonb_object_agg(
						o.id,
						jsonb_build_object(
							'has_image', o.image is not null,
							'image_spoilered', o.image_spoilered,
							'created_on', to_unix(o.created_on),
							'thread', o.thread,
							'body', o.body
						)
					) val
				from posts o
				where o.open
				group by o.thread
			) o on o.thread = t.id`,
		).
		Scan(&buf)
	if err != nil {
		return
	}
	if len(buf) == 0 {
		buf = []byte(`[]`)
	}
	return
}

// Retrieve public thread data from the database.
//
// page: page of the thread to fetch;
// 		 -1 to fetch the last page or;
// 		 -5 to fetch last 5 posts;
func GetThread(id uint64, page int) (thread []byte, err error) {
	err = db.
		QueryRow(
			context.Background(),
			"select get_thread($1, $2)",
			id,
			page,
		).
		Scan(&thread)
	castNoRows(&thread, &err)
	return
}

// Get all existing thread IDs
func GetThreadIDs() (ids []uint64, err error) {
	r, err := db.Query(context.Background(), "select id from threads")
	if err != nil {
		return
	}
	defer r.Close()

	for r.Next() {
		var id uint64
		err = r.Scan(&id)
		if err != nil {
			return
		}
		ids = append(ids, id)
	}
	err = r.Err()
	return
}

// Get the number of the last page of a thread
func GetLastPage(id uint64) (n int, err error) {
	err = db.
		QueryRow(
			context.Background(),
			`select coalesce(max(page), 0)
			from posts
			where thread = $1`,
			id,
		).
		Scan(&n)
	return
}

File: server/src/db/threads.rs
use super::{pool, PostInsertParams};
use crate::util::DynResult;
use common::payloads::{Post, ThreadWithPosts};

/// Parameters for inserting a thread and its OP
pub struct ThreadInsertParams<'a> {
	pub subject: &'a str,
	pub tags: &'a [String],
	pub op: PostInsertParams<'a>,
}

/// Insert thread and empty post into DB and return the thread ID
pub async fn insert_thread<'a>(
	p: &mut ThreadInsertParams<'a>,
) -> DynResult<u64> {
	let mut tx = pool().begin().await?;

	let id: i64 = sqlx::query!(
		"insert into threads (subject, tags)
		values ($1, $2)
		returning id",
		p.subject,
		p.tags,
	)
	.fetch_one(&mut tx)
	.await?
	.id;

	sqlx::query!(
		"insert into posts (
			id,
			thread,
			public_key,
			name,
			trip,
			flag,
			body
		)
		values (
			$1,
			$2,
			$3,
			$4,
			$5,
			$6,
			$7
		)",
		id,
		id,
		p.op.public_key.map(|i| i as i64),
		p.op.name,
		p.op.trip,
		p.op.flag,
		serde_json::to_value(p.op.body)?,
	)
	.execute(&mut tx)
	.await?;

	tx.commit().await?;

	Ok(id as u64)
}

/// Return all existing threads and their last 5 posts
pub async fn get_all_threads_short() -> DynResult<Vec<ThreadWithPosts>> {
	let mut threads: Vec<ThreadWithPosts> = Vec::new();
	let pool = pool();
	let mut s = sqlx::query!("select get_thread(id, -5) thread from threads")
		.fetch(&pool);
	while let Some(r) = s.next().await {
		if let Some(t) = r?.thread {
			threads.push(serde_json::from_value(t)?);
		}
	}

	Ok(threads)
}

/// Get a specific page of a thread
pub async fn get_page(thread: u64, page: u32) -> DynResult<Vec<Post>> {
	Ok(serde_json::from_value(
		sqlx::query!(
			"select jsonb_agg(encode(p)) page
			from posts p
			where thread = $1 and page = $2",
			thread as i64,
			page as i32,
		)
		.fetch_one(&pool())
		.await?
		.page
		.ok_or("query returned no JSON")?,
	)?)
}
